File: DragGesture-OC/DragGesture-OC/TestImageView.m
//
//  TestImageView.m
//  DragGesture-OC
//
//  Created by 杜奎 on 2019/4/28.
//  Copyright © 2019 du. All rights reserved.
//

#import "TestImageView.h"
#import "DragGestureHandler.h"

@interface TestImageView ()

@property (nonatomic, strong) DragGestureHandler *gesture;
@property (nonatomic, strong) UIView             *blackBgView;

@end

@implementation TestImageView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self addSubview:self.blackBgView];
        [self addSubview:self.imageView];
        
        __weak typeof(self) weakSelf = self;
        self.gesture = [[DragGestureHandler alloc] initWithGestureView:self.imageView bgView:self.blackBgView];
        self.gesture.completeBlock = ^(BOOL finish) {
            if (finish) {
                [weakSelf hide];
            }
        };
    }
    return self;
}

#pragma mark - action

- (void)show {
    [UIApplication.sharedApplication.keyWindow addSubview:self];
    self.blackBgView.alpha = 0;
    self.imageView.frame = self.outsideFrame;
    self.blackBgView.userInteractionEnabled = NO;
    [UIView animateWithDuration:0.3 animations:^{
        self.imageView.frame = self.insideFrame;
        self.blackBgView.alpha = 1.0;
    } completion:^(BOOL finished) {
        self.blackBgView.userInteractionEnabled = YES;
    }];
}

- (void)hide {
    self.blackBgView.userInteractionEnabled = NO;
    [UIView animateWithDuration:0.3 animations:^{
        self.imageView.frame = self.outsideFrame;
        self.blackBgView.alpha = 0;
    } completion:^(BOOL finished) {
        self.blackBgView.userInteractionEnabled = YES;
        [self removeFromSuperview];
    }];
}

- (void)tapAction {
    [self hide];
}

#pragma mark - setter & getter
- (UIView *)blackBgView {
    if (!_blackBgView) {
        _blackBgView = [[UIView alloc] initWithFrame:self.bounds];
        _blackBgView.backgroundColor = UIColor.blackColor;
        _blackBgView.userInteractionEnabled = NO;
        [_blackBgView addGestureRecognizer: [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapAction)]];
    }
    return _blackBgView;
}

- (UIImageView *)imageView {
    if (!_imageView) {
        _imageView = [[UIImageView alloc] init];
        _imageView.contentMode = UIViewContentModeScaleAspectFill;
        _imageView.clipsToBounds = YES;
    }
    return _imageView;
}

@end

File: DragGesture-Swift/DragGesture-Swift/TestImageView.swift
//
//  TestImageView.swift
//  DragGesture-Swift
//
//  Created by 杜奎 on 2019/4/27.
//  Copyright © 2019 du. All rights reserved.
//

import UIKit

class TestImageView: UIView {

    var outsideFrame = CGRect.zero
    var insideFrame = CGRect.zero
    var gesture: DragGestureHandler?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        self.addSubview(self.blackBgView)
        self.addSubview(self.imageView)
        
        self.gesture = DragGestureHandler.init(gestureView: self.imageView, bgView: self.blackBgView)
        self.gesture?.completeBlock = { [weak self] (finish) in
            if finish {
                self?.hide()
            }
        }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //MARK:- action
    
    func show() {
        UIApplication.shared.keyWindow?.addSubview(self)
        self.blackBgView.alpha = 0
        self.imageView.frame = self.outsideFrame
        self.blackBgView.isUserInteractionEnabled = false
        UIView.animate(withDuration: 0.3, animations: {
            self.imageView.frame = self.insideFrame
            self.blackBgView.alpha = 1
        }) { (finished) in
            self.blackBgView.isUserInteractionEnabled = true
        }
    }
    
    func hide() {
        self.blackBgView.isUserInteractionEnabled = false
        UIView.animate(withDuration: 0.3, animations: {
            self.imageView.frame = self.outsideFrame
            self.blackBgView.alpha = 0
        }) { (finished) in
            self.blackBgView.isUserInteractionEnabled = true
            self.removeFromSuperview()
        }
    }
    
    @objc func tapAction() {
        self.hide()
    }
    
    //MARK:- setter & getter
    
    private lazy var blackBgView: UIView = {
        let view = UIView.init(frame: self.bounds)
        view.backgroundColor = UIColor.black
        view.isUserInteractionEnabled = false
        view.addGestureRecognizer(UITapGestureRecognizer.init(target: self, action: #selector(tapAction)))
        return view
    }()
    
    private(set) var imageView: UIImageView = {
        let view = UIImageView.init()
        view.contentMode = .scaleAspectFill
        view.clipsToBounds = true
        return view
    }()
}
